tests/e2e/tests/pilot/performance/serviceentry_test.go

```diff
@@ -0,0 +1,328 @@
+			b = shot.Builder()
+
+			// wait for client to ACK the pushed snapshot
+			for {
+				if status := mcpServer.Cache.Status(snapshot.DefaultGroup); status != nil {
+					if status.LastWatchRequestTime().After(lastSnapshot) {
+						break
+					}
+				}
+				time.Sleep(10 * time.Millisecond)
+			}
+		}
+	}
+}
+
+func terminate() {
+	<-time.After(time.Second * time.Duration(*terminateAfter))
+}
+
+func adsConnectAndWait(n int, pilotAddr string) (adscs []*adsc.ADSC) {
+	for i := 0; i < n; i++ {
+		c, err := adsc.Dial(pilotAddr, "", &adsc.Config{
+			IP: net.IPv4(10, 10, byte(i/256), byte(i%256)).String(),
+		})
+		if err != nil {
+			log.Fatalf("Error connecting %s\n", err)
+		}
+		c.Watch()
+		_, err = c.Wait("rds", 10*time.Second)
+		if err != nil {
+			log.Fatalf("Error getting initial config %s\n", err)
+		}
+
+		if len(c.EDS) == 0 {
+			log.Fatal("No endpoints")
+		}
+		adscs = append(adscs, c)
+	}
+	return adscs
+}
+
+func runMcpServer() (*mcptest.Server, error) {
+	supportedTypes := make([]string, len(model.IstioConfigTypes))
+	for i, m := range model.IstioConfigTypes {
+		supportedTypes[i] = fmt.Sprintf("type.googleapis.com/%s", m.MessageName)
+	}
+	server, err := mcptest.NewServer(0, supportedTypes)
+	if err != nil {
+		return nil, err
+	}
+
+	return server, nil
+}
+
+func initLocalPilotTestEnv(t *testing.T, mcpPort int, grpcAddr, debugAddr string) util.TearDownFunc {
+	mixerEnv.NewTestSetup(mixerEnv.PilotMCPTest, t)
+	_, tearDown := util.EnsureTestServer(addMcpAddrs(mcpPort), setupPilotDiscoveryHTTPAddr(debugAddr), setupPilotDiscoveryGrpcAddr(grpcAddr))
+	return tearDown
+}
+
+func addMcpAddrs(mcpPort int) func(*bootstrap.PilotArgs) {
+	return func(arg *bootstrap.PilotArgs) {
+		arg.MCPServerAddrs = []string{fmt.Sprintf("mcp://127.0.0.1:%d", mcpPort)}
+	}
+}
+
+func setupPilotDiscoveryHTTPAddr(http string) func(*bootstrap.PilotArgs) {
+	return func(arg *bootstrap.PilotArgs) {
+		arg.DiscoveryOptions.HTTPAddr = http
+	}
+}
+
+func setupPilotDiscoveryGrpcAddr(grpc string) func(*bootstrap.PilotArgs) {
+	return func(arg *bootstrap.PilotArgs) {
+		arg.DiscoveryOptions.GrpcAddr = grpc
+	}
+}
+
+func registeredServiceEntries(apiEndpoint string) (int, error) {
+	resp, err := http.DefaultClient.Get(apiEndpoint)
+	if err != nil {
+		return 0, err
+	}
+	respBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return 0, err
+	}
+	debug := string(respBytes)
+	return strings.Count(debug, serviceDomain), nil
+}
+
+func generateServiceEntries() (envelopes []*mcp.Envelope) {
+	servicePorts := generatePorts(*serviceEntryCount)
+	backendPorts := generatePorts(*serviceEntryCount)
+	createTime := createTime()
+
+	for i := 0; i < *serviceEntryCount; i++ {
+		host := fmt.Sprintf("%s.%s", randName(), serviceDomain)
+		envelopes = append(envelopes, &mcp.Envelope{
+			Metadata: &mcp.Metadata{
+				Name:       fmt.Sprintf("serviceEntry-%s", randName()),
+				CreateTime: createTime,
+			},
+			Resource: &types.Any{
```

Comment:
Alternatively, you can use [types.MarshalAny](https://godoc.org/github.com/gogo/protobuf/types#MarshalAny) to encode the inner proto and set the TypeUrl.